lib/codecs.js

```diff
@@ -0,0 +1,105 @@
+'use strict';
+
+/* eslint-disable new-cap */
+
+var iconv = require('iconv-lite');
+var through = require('through2');
+
+var DEFAULT_ENCODING = require('./constants').DEFAULT_ENCODING;
+
+function Codec(codec, encoding) {
```

Comment:
I really like where this went with this abstraction but I think it can be improved upon.  I'm still formulating these ideas in my head.